File: script/reactive_navigation.py
#!/usr/bin/env python3

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan

class ReactiveNavigation:
    def __init__(self):  
        rospy.init_node('reactive_navigation', anonymous=True) 
        self.cmd_vel = Twist()
        self.robot_stopped = False
        self.obstacle_distance = 100.0
        self.laser_msg = None  
        self.laser_sub = rospy.Subscriber("base_scan", LaserScan, self.laser_cb, queue_size=1)
        self.cmd_vel_pub = rospy.Publisher("cmd_vel", Twist, queue_size=1)
        self.rate = rospy.Rate(5)

    def laser_cb(self, msg):
        self.laser_msg = msg
    
    def calculate_command(self):
        if self.laser_msg is None:
            return 
        
        if isinstance(self.laser_msg.ranges, list) or isinstance(self.laser_msg.ranges, tuple):

            self.obstacle_distance = min(self.laser_msg.ranges)
        
        if self.obstacle_distance > 1.0:
            self.cmd_vel.linear.x = 1.0
            self.cmd_vel.angular.z = 0.0

            
        else:
            self.cmd_vel.linear.x = 0.0
            self.cmd_vel.angular.z = 1.0

        self.cmd_vel_pub.publish(self.cmd_vel)
        self.obstacle_distance = 100.0  

    def run(self):
        while not rospy.is_shutdown():
            self.calculate_command()
            self.rate.sleep()

if __name__ == '__main__':
    try:
        controller = ReactiveNavigation()
        controller.run()
    except rospy.ROSInterruptException:
        pass

File: src/reactive_navigation.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <sensor_msgs/LaserScan.h>
#include <algorithm>
#include <vector>

class ReactiveNavigation {
public:
    ReactiveNavigation() : obstacle_distance(100.0), robot_stopped(false) {
        ros::NodeHandle nh;
        laser_sub = nh.subscribe("base_scan", 1, &ReactiveNavigation::laser_cb, this);
        cmd_vel_pub = nh.advertise<geometry_msgs::Twist>("cmd_vel", 1);
        rate = ros::Rate(5);  // 5 Hz
    }

    void laser_cb(const sensor_msgs::LaserScan::ConstPtr& msg) {
        laser_msg = msg;
    }

    void calculate_command() {
        if (!laser_msg) {
            return; 
        }

        obstacle_distance = *std::min_element(laser_msg->ranges.begin(), laser_msg->ranges.end());

        if (obstacle_distance > 0.6) {
            cmd_vel.linear.x = 1;
            cmd_vel.angular.z = 0.0;
        } else {
            cmd_vel.linear.x = 0.0;
            cmd_vel.angular.z = 1.0;
        }

        cmd_vel_pub.publish(cmd_vel);
        obstacle_distance = 100.0;  
    }

    void run() {
        while (ros::ok()) {
            calculate_command();
            rate.sleep();
        }
    }

private:
    geometry_msgs::Twist cmd_vel;
    bool robot_stopped;
    double obstacle_distance;
    sensor_msgs::LaserScan::ConstPtr laser_msg;
    ros::Subscriber laser_sub;
    ros::Publisher cmd_vel_pub;
    ros::Rate rate;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "reactive_navigation");
    try {
        ReactiveNavigation controller;
        controller.run();
    } catch (const ros::Exception& e) {
        ROS_ERROR("%s", e.what());
    }
    return 0;
}

